src/script/microbit-interfacing/MicrobitSerial.ts

```diff
@@ -31,68 +40,124 @@ class MicrobitSerial implements MicrobitConnection {
-
+    let unprocessedData = '';
     let previousButtonState = { A: 0, B: 0 };
-    await this.usb.startSerial(data => {
-      this.unprocessedInput += data;
-      let messages = protocol.splitMessages(this.unprocessedInput);
-      this.unprocessedInput = messages.remainingInput;
+
+    const processMessage = (data: string) => {
+      const messages = protocol.splitMessages(unprocessedData + data);
+      unprocessedData = messages.remainingInput;
       messages.messages.forEach(async msg => {
-        if (this.serialProtocolState === SerialProtocolState.AwaitingHandshakeResponse) {
-          let handshakeResponse = protocol.processHandshake(msg);
-          if (handshakeResponse && handshakeResponse.value === protocol.version) {
-            this.serialProtocolState = SerialProtocolState.Running;
-
-            // Request the micro:bit to start sending the periodic messages
-            const startCmd = protocol.generateCmdStart({
-              accelerometer: true,
-              buttons: true,
-            });
-            await this.usb.serialWrite(startCmd);
+        // Messages are either periodic sensor data or command/response
+        const sensorData = protocol.processPeriodicMessage(msg);
+        if (sensorData) {
+          inputBehaviour.accelerometerChange(
+            sensorData.accelerometerX,
+            sensorData.accelerometerY,
+            sensorData.accelerometerZ,
+          );
+          if (sensorData.buttonA !== previousButtonState.A) {
+            previousButtonState.A = sensorData.buttonA;
+            inputBehaviour.buttonChange(sensorData.buttonA, 'A');
           }
-        } else {
-          const sensorData = protocol.processPeriodicMessage(msg);
-          if (sensorData) {
-            inputBehaviour.accelerometerChange(
-              sensorData.accelerometerX,
-              sensorData.accelerometerY,
-              sensorData.accelerometerZ,
-            );
-            if (sensorData.buttonA !== previousButtonState.A) {
-              previousButtonState.A = sensorData.buttonA;
-              inputBehaviour.buttonChange(sensorData.buttonA, 'A');
-            }
-            if (sensorData.buttonB !== previousButtonState.B) {
-              previousButtonState.B = sensorData.buttonB;
-              inputBehaviour.buttonChange(sensorData.buttonB, 'B');
-            }
+          if (sensorData.buttonB !== previousButtonState.B) {
+            previousButtonState.B = sensorData.buttonB;
+            inputBehaviour.buttonChange(sensorData.buttonB, 'B');
           }
         }
+
+        const messageResponse = protocol.processResponseMessage(msg);
+        if (!messageResponse) {
+          return;
+        }
+        const responseResolve = this.responseMap.get(messageResponse.messageId);
+        if (responseResolve) {
+          this.responseMap.delete(messageResponse.messageId);
+          responseResolve(messageResponse);
+        }
```

Comment:
As we are struggling with performance, we could avoid doing this on every periodic message by either returning at the end of the `if (sensorData) {}` scope or only running this code if `sensorData` was `undefined`.

Reply:
Absolutely, thanks. 3dcd96245d2550c45c194e237c3b3d1f4785f774